typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt

```diff
@@ -350,63 +362,130 @@
                (list (syntax-property #'(void) 'mouse-over-tooltips (type-table->tooltips))))
   ;; report delayed errors
   (report-all-errors)
+  ;; provide-tbl : hash[id, listof[id]]
+  ;; maps internal names to all the names they're provided as
+  ;; XXX: should the external names be symbols instead of identifiers?
   (define provide-tbl
     (for/fold ([h (make-immutable-free-id-table)]) ([p (in-list provs)])
-      (define-syntax-class unknown-provide-form
-        (pattern
-         (~and name
-               (~or (~datum protect) (~datum for-syntax) (~datum for-label) (~datum for-meta)
-                    (~datum struct) (~datum all-from) (~datum all-from-except)
-                    (~datum all-defined) (~datum all-defined-except)
-                    (~datum prefix-all-defined) (~datum prefix-all-defined-except)
-                    (~datum expand)))))
       (syntax-parse p #:literal-sets (kernel-literals)
         [(#%provide form ...)
          (for/fold ([h h]) ([f (in-syntax #'(form ...))])
-           (parameterize ([current-orig-stx f])
-             (syntax-parse f
-               [i:id
-                (dict-update h #'i (lambda (tail) (cons #'i tail)) '())]
-               [((~datum rename) in out)
-                (dict-update h #'in (lambda (tail) (cons #'out tail)) '())]
-               [(name:unknown-provide-form . _)
-                (tc-error "provide: ~a not supported by Typed Racket" (syntax-e #'name.name))]
-               [_ (int-err "unknown provide form")])))]
+           (syntax-parse f
+             [i:id
+              (dict-update h #'i (lambda (tail) (cons #'i tail)) '())]
+             [((~datum rename) in out)
+              (dict-update h #'in (lambda (tail) (cons #'out tail)) '())]
+             [(name:unknown-provide-form . _)
```

Comment:
You mean rename the syntax class? There isn't anything currently called `unsupported-provide-form`.